contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+attempt to transition the current state of the system to the new declared 
+target state. 
+
+To facilitate update and rollback for these controllers, and to provide a 
+primitive that third party controllers can build on, we propose a mechanism 
+that allows controllers to manage a bounded history of revisions to the declared 
+target state of their generated Objects.
+
+## Affected Components
+
+1. API Machinery
+1. API Server
+1. Kubectl
+1. Controllers that utilize the feature
+
+## Requirements
+
+1. History is a collection of points in time, and each point in time must be 
+represented by its own Object. While it is tempting to aggregate all of an 
+Object's history into a single container Object, experience with Borg and Mesos 
+has taught us that this inevitably leads to exhausting the single Object size 
+limit of the system's the storage backend. 
+1. We must be able to select the Objects that contain point in time snapshots 
+of versions of an Object to reconstruct the Object's history.
+1. History respects causality. The Object type used to store point in time 
+snapshots must be strictly ordered with respect to creation. CreationTimestamp 
+should not be used, as this is susceptible to clock skew.
+1. History must not be revisionist. Once an Object corresponding to a version
+of a controllers target state is created, it can not be mutated.
+1. Controller history requires only current events. Storing an exhaustive 
+history of all revisions to all controllers is out of scope for our purposes, 
+and it can be solved by applying a version control system to manifests. Internal 
+revision history must only store revisions to the controller's target state that 
+correspond to live Objects and (potentially) a small, configurable number of 
+prior revisions.
+1. History is scale invariant. A revision to a controller is a modification 
+that changes the specification of the Objects it generates. Changing the 
+cardinality of those Objects is a scaling operation and does not constitute a 
+revision.
+
+## Terminology
+The following terminology is used throughout the rest of this proposal. We 
+make its meaning explicit here.
+- The specification type of a controller is the type that contains the 
+specification for the Objects generated by the controller.
+    - For example, the specification types for the ReplicaSet, DaemonSet, 
+     and StatefulSet controllers are ReplicaSetSpec, DaemonSetSpec, 
+     and StatefulSetSpec respectively.
+- The generated type(s) for a controller is/are the type of the Object(s) 
+generated by the controller. 
+    - Pod is a generated type for the ReplicaSet, DaemonSet, and StatefulSet 
+    controllers.
+    - PersistentVolumeClaim is also a generated type for the StatefulSet 
+    controller.
+- The current state of a controller is the union of the states of its generated 
+Objects along with its status.
+    - For ReplicaSet, DaemonSet, and StatefulSet, the current state of the 
```

Comment:
Since we are labeling the Pods to indicate their providence with respect to the revision of the controller's target state we are susceptible to downstream mutations by other controllers changing the controller's product. The best we can do is guarantee that our product meets the specification at the time of creation. If a third party mutates the product downstream (as long as it does so in a consistent and intentional way) we don't want to recall it and make it conform to the original spec.